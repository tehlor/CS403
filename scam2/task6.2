(println (powerSet '(y w e b u r)))
