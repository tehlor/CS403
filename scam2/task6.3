(println (powerSet '(z k t v u)))
