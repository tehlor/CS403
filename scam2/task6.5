(println (powerSet '(u r i p e w b)))
