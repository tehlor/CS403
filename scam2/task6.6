(println (powerSet '(l w k e)))
