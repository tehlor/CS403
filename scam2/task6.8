(println (powerSet '(u h g q a p t y l)))
