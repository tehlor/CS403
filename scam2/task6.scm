(define (main)
    (setNilDisplay 'nil)
    (setPort (open (getElement ScamArgs 1) 'read))
    (readArgs)
)

(define (readArgs)
    (define env this)
    (define (iter expr)
        (if (not (eof?))
            (begin (eval expr env) 
                (iter (readExpr))
            )
        )
    )
    (iter (readExpr))
)
;; let psetRest be all obtained subsets (ps (cdr set)) so we dont have to call it twice below 
;; append the list of all current subsets to the new subsets obtained by adding (car set) to all current subsets
;; cons car to all subsets of pset rest to obtain the new subsets
(define (F e T)
  (map
    (lambda (X)
      (cons e X)
      )
    T
    )
  )

(define (powerSet S)
  (cond
    ((null? S) '(()) )
    (else 
      (let ( (T (powerSet (cdr S))) (e (car S)) )
        (merge
          T
          (F e T)
          )
        )
      )
    )
  )

(define (merge a b)
  (define (help f s)
    (cond 
      ((null? f) s)
      ((null? s) f)
      (else
            (if (< (length (car f)) (length (car s)))
                (cons (car f) (help (cdr f) s))
                (cons (car s) (help f (cdr s)))
            )
        )
      )
    )
  (help a b)
)

