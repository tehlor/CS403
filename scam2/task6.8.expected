(nil (u) (h) (g) (q) (a) (p) (t) (y) (l) (u h) (u g) (u q) (u a) (u p) (u t) (u y) (u l) (h g) (h q) (h a) (h p) (h t) (h y) (h l) (g q) (g a) (g p) (g t) (g y) (g l) (q a) (q p) (q t) (q y) (q l) (a p) (a t) (a y) (a l) (p t) (p y) (p l) (t y) (t l) (y l) (u h g) (u h q) (u h a) (u h p) (u h t) (u h y) (u h l) (u g q) (u g a) (u g p) (u g t) (u g y) (u g l) (u q a) (u q p) (u q t) (u q y) (u q l) (u a p) (u a t) (u a y) (u a l) (u p t) (u p y) (u p l) (u t y) (u t l) (u y l) (h g q) (h g a) (h g p) (h g t) (h g y) (h g l) (h q a) (h q p) (h q t) (h q y) (h q l) (h a p) (h a t) (h a y) (h a l) (h p t) (h p y) (h p l) (h t y) (h t l) (h y l) (g q a) (g q p) (g q t) (g q y) (g q l) (g a p) (g a t) (g a y) (g a l) (g p t) (g p y) (g p l) (g t y) (g t l) (g y l) (q a p) (q a t) (q a y) (q a l) (q p t) (q p y) (q p l) (q t y) (q t l) (q y l) (a p t) (a p y) (a p l) (a t y) (a t l) (a y l) (p t y) (p t l) (p y l) (t y l) (u h g q) (u h g a) (u h g p) (u h g t) (u h g y) (u h g l) (u h q a) (u h q p) (u h q t) (u h q y) (u h q l) (u h a p) (u h a t) (u h a y) (u h a l) (u h p t) (u h p y) (u h p l) (u h t y) (u h t l) (u h y l) (u g q a) (u g q p) (u g q t) (u g q y) (u g q l) (u g a p) (u g a t) (u g a y) (u g a l) (u g p t) (u g p y) (u g p l) (u g t y) (u g t l) (u g y l) (u q a p) (u q a t) (u q a y) (u q a l) (u q p t) (u q p y) (u q p l) (u q t y) (u q t l) (u q y l) (u a p t) (u a p y) (u a p l) (u a t y) (u a t l) (u a y l) (u p t y) (u p t l) (u p y l) (u t y l) (h g q a) (h g q p) (h g q t) (h g q y) (h g q l) (h g a p) (h g a t) (h g a y) (h g a l) (h g p t) (h g p y) (h g p l) (h g t y) (h g t l) (h g y l) (h q a p) (h q a t) (h q a y) (h q a l) (h q p t) (h q p y) (h q p l) (h q t y) (h q t l) (h q y l) (h a p t) (h a p y) (h a p l) (h a t y) (h a t l) (h a y l) (h p t y) (h p t l) (h p y l) (h t y l) (g q a p) (g q a t) (g q a y) (g q a l) (g q p t) (g q p y) (g q p l) (g q t y) (g q t l) (g q y l) (g a p t) (g a p y) (g a p l) (g a t y) (g a t l) (g a y l) (g p t y) (g p t l) (g p y l) (g t y l) (q a p t) (q a p y) (q a p l) (q a t y) (q a t l) (q a y l) (q p t y) (q p t l) (q p y l) (q t y l) (a p t y) (a p t l) (a p y l) (a t y l) (p t y l) (u h g q a) (u h g q p) (u h g q t) (u h g q y) (u h g q l) (u h g a p) (u h g a t) (u h g a y) (u h g a l) (u h g p t) (u h g p y) (u h g p l) (u h g t y) (u h g t l) (u h g y l) (u h q a p) (u h q a t) (u h q a y) (u h q a l) (u h q p t) (u h q p y) (u h q p l) (u h q t y) (u h q t l) (u h q y l) (u h a p t) (u h a p y) (u h a p l) (u h a t y) (u h a t l) (u h a y l) (u h p t y) (u h p t l) (u h p y l) (u h t y l) (u g q a p) (u g q a t) (u g q a y) (u g q a l) (u g q p t) (u g q p y) (u g q p l) (u g q t y) (u g q t l) (u g q y l) (u g a p t) (u g a p y) (u g a p l) (u g a t y) (u g a t l) (u g a y l) (u g p t y) (u g p t l) (u g p y l) (u g t y l) (u q a p t) (u q a p y) (u q a p l) (u q a t y) (u q a t l) (u q a y l) (u q p t y) (u q p t l) (u q p y l) (u q t y l) (u a p t y) (u a p t l) (u a p y l) (u a t y l) (u p t y l) (h g q a p) (h g q a t) (h g q a y) (h g q a l) (h g q p t) (h g q p y) (h g q p l) (h g q t y) (h g q t l) (h g q y l) (h g a p t) (h g a p y) (h g a p l) (h g a t y) (h g a t l) (h g a y l) (h g p t y) (h g p t l) (h g p y l) (h g t y l) (h q a p t) (h q a p y) (h q a p l) (h q a t y) (h q a t l) (h q a y l) (h q p t y) (h q p t l) (h q p y l) (h q t y l) (h a p t y) (h a p t l) (h a p y l) (h a t y l) (h p t y l) (g q a p t) (g q a p y) (g q a p l) (g q a t y) (g q a t l) (g q a y l) (g q p t y) (g q p t l) (g q p y l) (g q t y l) (g a p t y) (g a p t l) (g a p y l) (g a t y l) (g p t y l) (q a p t y) (q a p t l) (q a p y l) (q a t y l) (q p t y l) (a p t y l) (u h g q a p) (u h g q a t) (u h g q a y) (u h g q a l) (u h g q p t) (u h g q p y) (u h g q p l) (u h g q t y) (u h g q t l) (u h g q y l) (u h g a p t) (u h g a p y) (u h g a p l) (u h g a t y) (u h g a t l) (u h g a y l) (u h g p t y) (u h g p t l) (u h g p y l) (u h g t y l) (u h q a p t) (u h q a p y) (u h q a p l) (u h q a t y) (u h q a t l) (u h q a y l) (u h q p t y) (u h q p t l) (u h q p y l) (u h q t y l) (u h a p t y) (u h a p t l) (u h a p y l) (u h a t y l) (u h p t y l) (u g q a p t) (u g q a p y) (u g q a p l) (u g q a t y) (u g q a t l) (u g q a y l) (u g q p t y) (u g q p t l) (u g q p y l) (u g q t y l) (u g a p t y) (u g a p t l) (u g a p y l) (u g a t y l) (u g p t y l) (u q a p t y) (u q a p t l) (u q a p y l) (u q a t y l) (u q p t y l) (u a p t y l) (h g q a p t) (h g q a p y) (h g q a p l) (h g q a t y) (h g q a t l) (h g q a y l) (h g q p t y) (h g q p t l) (h g q p y l) (h g q t y l) (h g a p t y) (h g a p t l) (h g a p y l) (h g a t y l) (h g p t y l) (h q a p t y) (h q a p t l) (h q a p y l) (h q a t y l) (h q p t y l) (h a p t y l) (g q a p t y) (g q a p t l) (g q a p y l) (g q a t y l) (g q p t y l) (g a p t y l) (q a p t y l) (u h g q a p t) (u h g q a p y) (u h g q a p l) (u h g q a t y) (u h g q a t l) (u h g q a y l) (u h g q p t y) (u h g q p t l) (u h g q p y l) (u h g q t y l) (u h g a p t y) (u h g a p t l) (u h g a p y l) (u h g a t y l) (u h g p t y l) (u h q a p t y) (u h q a p t l) (u h q a p y l) (u h q a t y l) (u h q p t y l) (u h a p t y l) (u g q a p t y) (u g q a p t l) (u g q a p y l) (u g q a t y l) (u g q p t y l) (u g a p t y l) (u q a p t y l) (h g q a p t y) (h g q a p t l) (h g q a p y l) (h g q a t y l) (h g q p t y l) (h g a p t y l) (h q a p t y l) (g q a p t y l) (u h g q a p t y) (u h g q a p t l) (u h g q a p y l) (u h g q a t y l) (u h g q p t y l) (u h g a p t y l) (u h q a p t y l) (u g q a p t y l) (h g q a p t y l) (u h g q a p t y l))
